src/mpid/ch4/netmod/ofi/ofi_rma.h

```diff
@@ -176,6 +176,136 @@ MPL_STATIC_INLINE_PREFIX void MPIDI_OFI_sigreq_complete(MPIR_Request ** sigreq)
     }
 }
 
+enum {
+    MPIDI_OFI_PUT,
+    MPIDI_OFI_GET,
+};
+
+static inline int MPIDI_OFI_nopack_putget(const void *origin_addr, int origin_count,
+                                          MPI_Datatype origin_datatype, int target_rank,
+                                          MPI_Aint target_disp, int target_count,
+                                          MPI_Datatype target_datatype, MPIR_Win * win,
+                                          MPIDI_av_entry_t * addr, int rma_type,
+                                          MPIR_Request ** sigreq)
+{
+    int mpi_errno = MPI_SUCCESS;
+    uint64_t flags;
+    struct fi_msg_rma msg;
+    struct fi_rma_iov riov;
+    struct iovec iov;
+    size_t target_bytes, origin_bytes;
+
+    MPIR_Datatype_get_size_macro(origin_datatype, origin_bytes);
+    origin_bytes *= origin_count;
+    MPIR_Datatype_get_size_macro(target_datatype, target_bytes);
+    target_bytes *= target_count;
+
+    /* allocate request */
+    MPIDI_OFI_win_request_t *req = MPIDI_OFI_win_request_create();
+    MPIR_ERR_CHKANDSTMT((req) == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail, "**nomemreq");
+    req->event_id = MPIDI_OFI_EVENT_ABORT;
+    req->next = MPIDI_OFI_WIN(win).syncQ;
+    MPIDI_OFI_WIN(win).syncQ = req;
+
+    /* populate target iovecs */
+    struct iovec *target_iov;
+    void *target_base;
+    MPI_Aint total_target_iov_len;
+    MPI_Aint target_iov_offset = 0;
+    MPI_Aint max_iov_len;
+    MPI_Aint target_iov_len;
+    MPIR_Typerep_iov_len(target_count, target_datatype, target_bytes, &total_target_iov_len);
+    max_iov_len = MPL_MIN(total_target_iov_len, MPIR_CVAR_CH4_OFI_RMA_IOVEC_MAX);
+    target_iov = MPL_malloc(sizeof(struct iovec) * max_iov_len, MPL_MEM_RMA);
+    target_base =
+        (void *) (MPIDI_OFI_winfo_base(win, target_rank) +
+                  (target_disp * MPIDI_OFI_winfo_disp_unit(win, target_rank)));
+    MPIR_Typerep_to_iov_offset(target_base, target_count, target_datatype, target_iov_offset,
+                               target_iov, max_iov_len, &target_iov_len);
```

Comment:
I'm confused here.  Doesn't this need an extra loop to refill the IOV vector and issue the operations for cases where `total_target_iov_len` > `MPIR_CVAR_CH4_OFI_RMA_IOVEC_MAX`.

Reply:
For testing purposes, it'll be good to add a hack patch that temporarily sets the default IOV length to a smaller number (e.g., 1024), so that path can be tested too.  We'd obviously change it back to 16384 for the actual merge.

Reply:
Yes, that happens in the while loop below. These could probably be unified to avoid the duplication.

Reply:
Oh, I see.  Yeah, it's a bit confusing to have one IOV load here and the rest down in the loop.  It might be better to have it in one place.